app/src/main/java/org/schabi/newpipe/local/bookmark/BookmarkFragment.java

```diff
@@ -247,12 +314,242 @@ protected void resetFragment() {
+                    + "is not loaded or playlists not modified");
+            return;
+        }
+
+        final List<LocalItem> items = itemListAdapter.getItemsList();
+        final List<PlaylistMetadataEntry> localItemsUpdate = new ArrayList<>();
+        final List<Long> localItemsDeleteUid = new ArrayList<>();
+        final List<PlaylistRemoteEntity> remoteItemsUpdate = new ArrayList<>();
+        final List<Long> remoteItemsDeleteUid = new ArrayList<>();
+
+        // Calculate display index
+        for (int i = 0; i < items.size(); i++) {
+            final LocalItem item = items.get(i);
+
+            if (item instanceof PlaylistMetadataEntry) {
+                ((PlaylistMetadataEntry) item).displayIndex = i;
+
+                final Long uid = ((PlaylistMetadataEntry) item).uid;
+                final Pair<Long, LocalItem.LocalItemType> key = new Pair<>(uid,
+                        LocalItem.LocalItemType.PLAYLIST_LOCAL_ITEM);
+                final Long databaseIndex = displayIndexInDatabase.remove(key);
+
+                // The database index should not be null because inserting new item into database
+                // is not handled here. NullPointerException has occurred once, but I can't
+                // reproduce it. Enhance robustness here.
+                if (databaseIndex != null && databaseIndex != i) {
+                    localItemsUpdate.add((PlaylistMetadataEntry) item);
+                }
+            } else if (item instanceof PlaylistRemoteEntity) {
+                ((PlaylistRemoteEntity) item).setDisplayIndex(i);
+
+                final Long uid = ((PlaylistRemoteEntity) item).getUid();
+                final Pair<Long, LocalItem.LocalItemType> key = new Pair<>(uid,
+                        LocalItem.LocalItemType.PLAYLIST_REMOTE_ITEM);
+                final Long databaseIndex = displayIndexInDatabase.remove(key);
+
+                if (databaseIndex != null && databaseIndex != i) {
+                    remoteItemsUpdate.add((PlaylistRemoteEntity) item);
+                }
+            }
+        }
+
+        // Find deleted items
+        for (final Pair<Long, LocalItem.LocalItemType> key : displayIndexInDatabase.keySet()) {
+            if (key.second.equals(LocalItem.LocalItemType.PLAYLIST_LOCAL_ITEM)) {
+                localItemsDeleteUid.add(key.first);
+            } else if (key.second.equals(LocalItem.LocalItemType.PLAYLIST_REMOTE_ITEM)) {
+                remoteItemsDeleteUid.add(key.first);
+            }
+        }
+
+        displayIndexInDatabase.clear();
+
+        // 1. Update local playlists
+        // 2. Update remote playlists
+        // 3. Set isModified false
+        disposables.add(localPlaylistManager.updatePlaylists(localItemsUpdate, localItemsDeleteUid)
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(() -> disposables.add(remotePlaylistManager.updatePlaylists(
+                        remoteItemsUpdate, remoteItemsDeleteUid)
```

Comment:
Instead of nesting things, you can just use some RxJava functions to do two actions in parallel and then calling `debounceSaver.setIsModified(false);` when both have finished.